br/pkg/lightning/backend/external/iter.go

```diff
@@ -0,0 +1,385 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package external
+
+import (
+	"bytes"
+	"container/heap"
+	"context"
+	"io"
+
+	"github.com/pingcap/tidb/br/pkg/storage"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+)
+
+type heapElem interface {
+	sortKey() []byte
+}
+
+type sortedReader[T heapElem] interface {
+	path() string
+	next() (T, error)
+	close() error
+}
+
+type mergeHeapElem[T heapElem] struct {
+	elem      T
+	readerIdx int
+}
+
+type mergeHeap[T heapElem] []mergeHeapElem[T]
+
+func (h mergeHeap[T]) Len() int {
+	return len(h)
+}
+
+func (h mergeHeap[T]) Less(i, j int) bool {
+	return bytes.Compare(h[i].elem.sortKey(), h[j].elem.sortKey()) < 0
+}
+
+func (h mergeHeap[T]) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *mergeHeap[T]) Push(x interface{}) {
+	*h = append(*h, x.(mergeHeapElem[T]))
+}
+
+func (h *mergeHeap[T]) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+type mergeIter[T heapElem, R sortedReader[T]] struct {
+	h             mergeHeap[T]
+	readers       []*R
+	curr          T
+	lastReaderIdx int
+	err           error
+
+	logger *zap.Logger
+}
+
+// newMergeIter creates a merge iterator for multiple sorted readers. the
+// ownership of readers is transferred to the mergeIter. When newMergeIter
+// returns error, the reader will be closed.
+// To reduce duplication, we tolerate caller passing nil readers which means
+// reader's source file is empty.
+func newMergeIter[
+	T heapElem,
+	R sortedReader[T],
+](readers []*R, logger *zap.Logger) (*mergeIter[T, R], error) {
+	i := &mergeIter[T, R]{
+		h:             make(mergeHeap[T], 0, len(readers)),
+		readers:       readers,
+		lastReaderIdx: -1,
+		logger:        logger,
+	}
+	for j := range i.readers {
+		if i.readers[j] == nil {
+			continue
+		}
+		rd := *i.readers[j]
+		e, err := rd.next()
+		if err == io.EOF {
+			closeErr := rd.close()
+			if closeErr != nil {
+				i.logger.Warn("failed to close reader",
+					zap.String("path", rd.path()),
+					zap.Error(closeErr))
+			}
+			i.readers[j] = nil
+			continue
+		}
+		if err != nil {
+			i.close()
+			return nil, err
+		}
+		i.h = append(i.h, mergeHeapElem[T]{
+			elem:      e,
+			readerIdx: j,
+		})
+	}
+	heap.Init(&i.h)
+	return i, nil
+}
+
+// close must be called for a mergeIter when it is no longer used.
+func (i *mergeIter[T, R]) close() error {
+	var firstErr error
+	for idx, rdp := range i.readers {
+		if rdp != nil {
+			rd := *rdp
+			err := rd.close()
```

Comment:
Maybe `rdp.close()` directly?

Reply:
golang can't do it. Seems the methods set rule is no longer applied to generic type.